src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs

```diff
@@ -189,12 +187,46 @@ private static string GetTypeGuid(MetadataReader reader, TypeDefinition type)
                 if (reader.StringComparer.Equals(attributeType.Namespace, "System.Runtime.InteropServices") && reader.StringComparer.Equals(attributeType.Name, "GuidAttribute"))
                 {
                     CustomAttributeValue<KnownType> data = attribute.DecodeValue(new TypeResolver());
-                    return (string)data.FixedArguments[0].Value;
+                    return $"{{{(string)data.FixedArguments[0].Value}}}";
                 }
             }
             throw new BuildErrorException(Strings.ClsidMapExportedTypesRequireExplicitGuid, GetTypeName(reader, type));
         }
 
+        private static void WriteClsidMap(Stream clsidMapFile, Dictionary<string, ClsidEntry> map)
+        {
+            const int SyncWriteThreshold = 1_000; // Copy to stream after writing this many bytes
```

Comment:
I find it quite troubling that we're expecting every consumer to think at this level. The BCL should provide a way to write Json to a Stream, IMHO.

Reply:
Where does 1 MB guideline for synchronous case come from? How would we know to choose that?

Reply:
(If we were writing a big payload)

Reply:
The 1 MB guidelines is an implementation detail of using ArrayBufferWriter which is backed by the array pool. You can definitely continue to write up to 2 GB until OOM, but beyond 1 MB, the array pool starts to allocate. That's why I suggested, that IF you were to set a sync threshold, 1 MB is reasonable. You don't have to, though, especially if allocations aren't a concern (writing to a growable stream without a sync threshold, for instance memory stream, has the same concern).

Reply:
That's fine if you need the control, but as you spread this to every consumer of NewtonSoft.Json in the stack, you're going to hit many that don't. Not being able to just write Json to a file is very odd for many of these components.

Reply:
>  (writing to a growable stream without a sync threshold, for instance memory stream, has the same concern).

I would expect to be able to switch to FileStream from MemoryStream in that case, not start worrying about buffering myself.

Reply:
That's reasonable feedback. We made an deliberate trade-off between usability and performance. Writing to a (file)stream is definitely a common scenario, and if I can make it easier without having the lowest layer of the JSON stack (i.e. JsonWriter) regress in performance, then I would. However, since our primary focus was not high usability, but rather an emphasis on performance, the design required synchronous buffered writes (to an IBufferWriter) with periodic streamed-writes to be done by the caller (which I don't think we can change at this layer). FWIW, the serializer layer, built on top of the writer, let's you write to a stream.

Reply:
It feels like there should be a thing between serialization and this layer that lets me write forward-only to a Stream. I don't get how that would regress the lower layer, which someone could still use if they really need it.

Reply:
We could certainly consider adding such a type in the future. Though, it would likely contain much of the same API surface as the lowest layer (overloads like WriteString, WriteNumber, etc.), but those would just end up writing directly to the stream rather than an internal buffer. The caller of the writer is actively calling the write methods today, so I don't necessarily see periodically flushing to a stream (by tracking the `BytesWritten` property) as such a significant pain-point, to justify the feature at the moment, tbh. But, I am open to the feedback, and will consider it. Thanks.

Reply:
Could terre be `StreamBufferWriter : IBufferWriter`?

Reply:
Maybe what I really want are the higher level things like serialization or DOM and not a stream binding for forward-only writing. It's surprising that this doesn't exist, but it's probably still lower than what we'd use if there were a full JSON stack available.

System.Reflection.Metadata (which I maintained) has similar tradeoffs. It is a low level API by design upon which higher levels can be built. I often told people that it was not the direct API for their higher level scenarios. Maybe it's premature to push tooling away from NewtonSoft.Json without the higher level parts being implemented. This is not the sort of code I expect to maintain at this level of the stack. I fully appreciate that there are scenarios for dropping down this far, but I don't think the json files read and written by build tooling are a good fit.

Reply:
Maybe what I really want are the higher level things like serialization or DOM and not a stream binding for forward-only writing. It's surprising that this doesn't exist, but it's probably still lower than what we'd use if there were a full JSON stack available.

System.Reflection.Metadata (which I maintained) has similar tradeoffs. It is a low level API by design upon which higher levels can be built. I often told people that it was not the direct API for their higher level scenarios. Maybe it's premature to push tooling away from NewtonSoft.Json without the higher level parts being implemented. This is not the sort of code I expect to maintain at this level of the stack. I fully appreciate that there are scenarios for dropping down this far, but I don't think the json files read and written by build tooling are a good fit.

Reply:
> Maybe what I really want are the higher level things like serialization or DOM and not a stream binding for forward-only writing. It's surprising that this doesn't exist

I am not sure, I follow. It does exist, with plans to build up the feature capabilities:
[JsonSerializer](https://github.com/dotnet/corefx/blob/6c027aebb8fd1c643401bf801083cff5e6d34853/src/System.Text.Json/ref/System.Text.Json.cs#L313-L327)
[JsonDocument](https://github.com/dotnet/corefx/blob/6c027aebb8fd1c643401bf801083cff5e6d34853/src/System.Text.Json/ref/System.Text.Json.cs#L16-L29)

Reply:
The source package (or at least version 4.6.0-preview3.19128.7 which I pulled from NuGet.org) doesn't contain the `JsonSerializer` class. Looks like the newer versions are all on the blob feed though so we could pull down a newer version with the serialization support.

Reply:
Let me rephrase. I wassaying "maybe I really want Serializer or DOM, but I am surprised that forward-only-writer-to-stream is not directly supported".

JsonDocument is read-only today AFAICT from above. Are there plans to be able to build one up programatically and save like XDocument?

(I reach for XDocument a lot more than Xml serialization because I've been burned by serialization many, many times.)

Reply:
I did think that there was no JsonSerializer avaialble yet though, happy to see that is in.

Reply:
> Are there plans to be able to build one up programatically and save like XDocument?

Not for 3.0, but something we are considering adding in the future based on feedback/user scenarios like yours.

Reply:
FWIW, your concerns and feedback have been heard and hopefully addressed with the recent `Utf8JsonWriter` changes! We ended up making some trade-off between perf and usability to enable the scenarios you mentioned, @nguerrera :)

https://gist.github.com/ahsonkhan/f6f30656717548212693e5eaa49cece5
https://github.com/dotnet/corefx/pull/36961

Presumably, we should re-visit this PR given the new changes!

Reply:
Sweet! Thanks!